src/platform/web/ui/general/LazyListView.js

```diff
@@ -0,0 +1,250 @@
+import {el} from "./html.js";
+import {mountView} from "./utils.js";
+import {insertAt, ListView} from "./ListView.js";
+
+class ItemRange {
+    constructor(topCount, renderCount, bottomCount) {
+        this.topCount = topCount;
+        this.renderCount = renderCount;
+        this.bottomCount = bottomCount;
+    }
+
+    contains(range) {
+        // don't contain empty ranges
+        // as it will prevent clearing the list
+        // once it is scrolled far enough out of view
+        if (!range.renderCount && this.renderCount) {
+            return false;
+        }
+        return range.topCount >= this.topCount &&
+            (range.topCount + range.renderCount) <= (this.topCount + this.renderCount);
+    }
+
+    containsIndex(idx) {
+        return idx >= this.topCount && idx <= (this.topCount + this.renderCount);
+    }
+
+    expand(amount) {
+        // don't expand ranges that won't render anything
+        if (this.renderCount === 0) {
+            return this;
+        }
+
+        const topGrow = Math.min(amount, this.topCount);
+        const bottomGrow = Math.min(amount, this.bottomCount);
+        return new ItemRange(
+            this.topCount - topGrow,
+            this.renderCount + topGrow + bottomGrow,
+            this.bottomCount - bottomGrow,
+        );
+    }
+
+    totalSize() {
+        return this.topCount + this.renderCount + this.bottomCount;
+    }
+
+    normalize(idx) {
+        /*
+        map index from list to index in rendered range
+        eg: if the index range of this._list is [0, 200] and we have rendered
+        elements in range [50, 60] then index 50 in list must map to index 0
+        in DOM tree/childInstance array.
+        */
+        return idx - this.topCount;
+    }
+}
+
+export class LazyListView extends ListView {
```

Comment:
I wonder how behaviour is still used from `ListView`? Do you think it could be simpler to just implement this from scratch?

Reply:
Very little. It would me much simpler to implement this from scratch indeed. You can see in some places (like `loadList`) where the code gets a little crufty because we have to support the super class.